### PR TITLE
Cleaning up the middleware-based API implementation.

### PR DESCRIPTION
**Pull request by halostatue**

### CODE REVIEW

**Comment by halostatue** [comment]
As noted on pandurang90/cloudconvert#9, this cleans up some errors and inconsistencies I saw in the code you proposed to @pandurang90 for phase 2.

**Comment by olibob** [comment]
@halostatue thanks, I'll look into it today.

**Comment by olibob** [comment]
**Configuration** class

I understand the CONVERSION_URL and validation.@pandurang90 's configuration code is a pretty standard configuration pattern. I don't see why this needs to change.

**Client** class

Once set, the configuration is accessible anywhere. Adding the configuration as an initialisation parameter seems very weird to me. Even if you'd want to do this, a validation in Upload is not necessary as nothing can happen without a new_process (where the validation will be done). The only way this makes sense is using different configurations on the fly.

**Conversion** class

Adding an `options = {}` parameter to the initialisation method for the configuration makes sense only if you are going to work with several different configurations. What use case do you see for it?

File: Gemfile [hunk 1]
```diff
@@ -2,6 +2,3 @@ source 'https://rubygems.org'
 
 # Specify your gem's dependencies in cloudconvert.gemspec
 gemspec
```
Comment (olibob) [inline Gemfile:4]: Yes

File: lib/cloudconvert/client.rb [hunk 2]
```diff
@@ -23,14 +22,14 @@ def get(url)
...
-    def initialize
-
-      @up ||= Faraday.new(url: Cloudconvert::CONVERSION_URL) do |faraday|
-        faraday.request 			:multipart
-        faraday.request 			:url_encoded
-        faraday.response			:json, content_type: /\bjson$/
-        # faraday.response			:logger
-        faraday.adapter 			Faraday.default_adapter
+    def initialize(config = Cloudconvert.configuration)
+      config.validate!
```
Comment (olibob) [inline lib/cloudconvert/client.rb]: useless, done in Client.initialize

File: lib/cloudconvert/configuration.rb [hunk 3]
```diff
@@ -1,21 +1,32 @@
...
     def initialize
-      api_key = nil
-      callback = nil
+      @api_key  = nil
+      @callback = nil
+      @api_url  = CONVERSION_URL
+    end
+
+    def validate!
+      raise MissingAPIKey if api_key.nil?
```
Comment (olibob) [inline lib/cloudconvert/configuration.rb:16]: cleaner, yes

File: lib/cloudconvert/configuration.rb [hunk 4]
```diff
@@ -1,21 +1,32 @@
...
     end
-    
   end
 
   class << self
```
Comment (olibob) [inline lib/cloudconvert/configuration.rb:20]: I don't see why this needs to change

File: lib/cloudconvert/conversion.rb [hunk 5]
```diff
@@ -1,101 +1,105 @@
 module Cloudconvert
```
Comment (olibob) [inline lib/cloudconvert/conversion.rb:1]: Make sure to your pull request reflects changes in code, not in indentations or spaces.

**Comment by halostatue** [comment]
This is a minimal set of changes that I've made—over the last ~12 hours, I've been doing a _substantial_ change that I think will make things even better; it's built off of this and a few other things. I'll be submitting that as a PR to @pandurang90’s repo directly once I've got it, because it has much larger implications (think DSL-driven instead of hash-driven) and makes for a much more fluent API.

To address your comments:
1. _Configuration_ has been made reusable/resettable because while the configuration pattern is pretty standard, it's also broken for a whole bunch of use cases (multi-tenant applications are a good case where it breaks). The way that I've written this up, the behaviour is completely transparent to people who only use the global configuration, but it gives the flexibility for alternative configurations.
2. In my larger change, I'm moving from the _Conversion_ class to a _Process_ class and am working on subsuming the the current `new_process` behaviour to be fairly automatic.
3. It's a small change to not use revalidate configuration on Upload, and you're right about that. Passing the configuration as an upload parameter is necessary to fix point 1 (make configuration passable so multi-tenant applications could work with this).
4. The PR explicitly includes some formatting changes because there were mixed tabs and spaces and formatting inconsistent with fairly well-established Ruby convention. The easiest way to see this cleanly is to add `?w=1` to the end of the [files link](https://github.com/olibob/cloudconvert/pull/1/files?w=1); at that point, only the code changes are visible.

Like I said, I've got a much larger proposed change coming that's based on your work and this cleanup from that. It represents a bit of a change in spirit, but it will be _cleaner_ to work with. Think something like:

``` ruby
Cloudconvert.configure do |config|
  config.api_key = 'api-key'
end

converter = Cloudconvert.convert do |process|
  process.input_format 'doc'
  process.output_format 'pdf'
  process.from_s3 { access_key_id: 'keyid', secret_key_id: 'sekret', bucket: 'bucketname' }
  process.file 'resume.doc'
  process.dropbox
end

converter.run do |status, process|
  …
end

puts converter.download_link if converter.successful?
```

You can still do the asynchronous mechanism, but you can use either `#run` or `#wait` to synchronize, the way you wrote in the README.

**Comment by olibob** [comment]
Interesting, I'm looking forward to see what you come up with. Sounds great.

**Comment by halostatue** [comment]
I've made a fluent API PR at pandurang90/cloudconvert#10. I recommend that it be merged instead of this (it incorporates these changes), as it has a more fluent API and I think represents a good basis. It wouldn't have been possible without the work you did in this and the existence of the library in the first place.

**Pull request closed without merging**